File: runtime/process.py
"""Terminal process management for agent subprocesses."""

from __future__ import annotations

import logging
import os
import subprocess
import threading
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Optional, Callable, Dict, List

logger = logging.getLogger(__name__)


@dataclass
class AgentProcess:
    """Represents a running agent subprocess."""
    agent_id: str
    process: subprocess.Popen
    working_dir: Path
    started_at: datetime = field(default_factory=datetime.now)
    output_lines: List[str] = field(default_factory=list)
    error_lines: List[str] = field(default_factory=list)

    @property
    def pid(self) -> Optional[int]:
        return self.process.pid if self.process else None

    @property
    def is_running(self) -> bool:
        return self.process is not None and self.process.poll() is None

    @property
    def return_code(self) -> Optional[int]:
        if self.process:
            return self.process.poll()
        return None

    def send_input(self, text: str):
        """Send input to the process stdin."""
        if self.process and self.process.stdin:
            try:
                self.process.stdin.write(text + "\n")
                self.process.stdin.flush()
            except (BrokenPipeError, OSError) as e:
                logger.error(f"Failed to send input to {self.agent_id}: {e}")

    def terminate(self, timeout: float = 5.0):
        """Gracefully terminate the process."""
        if not self.process:
            return

        try:
            self.process.terminate()
            self.process.wait(timeout=timeout)
        except subprocess.TimeoutExpired:
            logger.warning(f"Process {self.agent_id} didn't terminate, killing")
            self.process.kill()
            self.process.wait(timeout=2.0)

    def kill(self):
        """Forcefully kill the process."""
        if self.process:
            self.process.kill()
            self.process.wait(timeout=2.0)


class TerminalManager:
    """Manages terminal processes for agents."""

    def __init__(self, base_dir: Path):
        self.base_dir = Path(base_dir)
        self._processes: Dict[str, AgentProcess] = {}
        self._output_threads: Dict[str, threading.Thread] = {}
        self._callbacks: Dict[str, List[Callable[[str, str], None]]] = {}
        self._lock = threading.Lock()

    def spawn_claude_agent(
        self,
        agent_id: str,
        prompt: str,
        working_dir: Optional[Path] = None,
        timeout: int = 300,
        dangerously_skip_permissions: bool = True,
    ) -> AgentProcess:
        """
        Spawn a Claude Code CLI agent.

        Args:
            agent_id: Unique identifier for this agent
            prompt: The initial prompt for Claude
            working_dir: Working directory for the agent
            timeout: Timeout in seconds
            dangerously_skip_permissions: Skip permission prompts

        Returns:
            AgentProcess instance
        """
        working_dir = working_dir or self.base_dir / "sandbox"
        working_dir = Path(working_dir)
        working_dir.mkdir(parents=True, exist_ok=True)

        cmd = ["claude", "--print"]
        if dangerously_skip_permissions:
            cmd.append("--dangerously-skip-permissions")
        cmd.extend(["-p", prompt])

        logger.info(f"Spawning agent {agent_id} in {working_dir}")

        process = subprocess.Popen(
            cmd,
            cwd=str(working_dir),
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            stdin=subprocess.PIPE,
            text=True,
            bufsize=1,
        )

        agent_process = AgentProcess(
            agent_id=agent_id,
            process=process,
            working_dir=working_dir,
        )

        with self._lock:
            self._processes[agent_id] = agent_process

        # Start output reader thread
        self._start_output_reader(agent_id, agent_process)

        return agent_process

    def spawn_shell(
        self,
        agent_id: str,
        command: str,
        working_dir: Optional[Path] = None,
        env: Optional[Dict[str, str]] = None,
    ) -> AgentProcess:
        """
        Spawn a shell process for an agent.

        Args:
            agent_id: Unique identifier for this agent
            command: The shell command to run
            working_dir: Working directory
            env: Environment variables

        Returns:
            AgentProcess instance
        """
        working_dir = working_dir or self.base_dir / "sandbox"
        working_dir = Path(working_dir)
        working_dir.mkdir(parents=True, exist_ok=True)

        process_env = os.environ.copy()
        if env:
            process_env.update(env)

        logger.info(f"Spawning shell for {agent_id}: {command}")

        process = subprocess.Popen(
            command,
            shell=True,
            cwd=str(working_dir),
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            stdin=subprocess.PIPE,
            text=True,
            bufsize=1,
            env=process_env,
        )

        agent_process = AgentProcess(
            agent_id=agent_id,
            process=process,
            working_dir=working_dir,
        )

        with self._lock:
            self._processes[agent_id] = agent_process

        self._start_output_reader(agent_id, agent_process)

        return agent_process

    def _start_output_reader(self, agent_id: str, agent_process: AgentProcess):
        """Start a thread to read process output."""
        def reader():
            while agent_process.is_running:
                try:
                    line = agent_process.process.stdout.readline()
                    if line:
                        agent_process.output_lines.append(line.rstrip())
                        self._notify_output(agent_id, "stdout", line.rstrip())
                except Exception as e:
                    logger.error(f"Error reading stdout for {agent_id}: {e}")
                    break

            # Read any remaining stderr
            try:
                stderr = agent_process.process.stderr.read()
                if stderr:
                    for line in stderr.splitlines():
                        agent_process.error_lines.append(line)
                        self._notify_output(agent_id, "stderr", line)
            except Exception as e:
                logger.error(f"Error reading stderr for {agent_id}: {e}")

        thread = threading.Thread(target=reader, daemon=True)
        thread.start()

        with self._lock:
            self._output_threads[agent_id] = thread

    def _notify_output(self, agent_id: str, stream: str, line: str):
        """Notify registered callbacks of output."""
        callbacks = self._callbacks.get(agent_id, [])
        for callback in callbacks:
            try:
                callback(stream, line)
            except Exception as e:
                logger.error(f"Callback error for {agent_id}: {e}")

    def register_output_callback(
        self,
        agent_id: str,
        callback: Callable[[str, str], None]
    ):
        """Register a callback for agent output."""
        with self._lock:
            if agent_id not in self._callbacks:
                self._callbacks[agent_id] = []
            self._callbacks[agent_id].append(callback)

    def get_process(self, agent_id: str) -> Optional[AgentProcess]:
        """Get an agent's process."""
        with self._lock:
            return self._processes.get(agent_id)

    def get_output(self, agent_id: str) -> List[str]:
        """Get all stdout output for an agent."""
        process = self.get_process(agent_id)
        return process.output_lines if process else []

    def get_errors(self, agent_id: str) -> List[str]:
        """Get all stderr output for an agent."""
        process = self.get_process(agent_id)
        return process.error_lines if process else []

    def is_running(self, agent_id: str) -> bool:
        """Check if an agent's process is still running."""
        process = self.get_process(agent_id)
        return process.is_running if process else False

    def wait_for_completion(
        self,
        agent_id: str,
        timeout: Optional[float] = None
    ) -> Optional[int]:
        """
        Wait for an agent's process to complete.

        Args:
            agent_id: The agent ID
            timeout: Timeout in seconds

        Returns:
            Return code, or None if timeout
        """
        process = self.get_process(agent_id)
        if not process or not process.process:
            return None

        try:
            return process.process.wait(timeout=timeout)
        except subprocess.TimeoutExpired:
            return None

    def terminate(self, agent_id: str, timeout: float = 5.0):
        """Terminate an agent's process."""
        process = self.get_process(agent_id)
        if process:
            process.terminate(timeout)
            with self._lock:
                if agent_id in self._processes:
                    del self._processes[agent_id]
                if agent_id in self._output_threads:
                    del self._output_threads[agent_id]

    def terminate_all(self, timeout: float = 5.0):
        """Terminate all agent processes."""
        with self._lock:
            agent_ids = list(self._processes.keys())

        for agent_id in agent_ids:
            self.terminate(agent_id, timeout)

    def get_all_agents(self) -> List[str]:
        """Get list of all agent IDs."""
        with self._lock:
            return list(self._processes.keys())

    def get_running_agents(self) -> List[str]:
        """Get list of running agent IDs."""
        with self._lock:
            return [
                agent_id for agent_id, process in self._processes.items()
                if process.is_running
            ]

File: src/runtime/process.js
/**
 * @file Terminal process management for agent subprocesses.
 * @module runtime/process
 */

import { spawn } from 'node:child_process';
import { mkdir } from 'node:fs/promises';
import { resolve } from 'node:path';
import { EventEmitter } from 'node:events';

/**
 * Represents a running agent subprocess.
 * @extends EventEmitter
 */
export class AgentProcess extends EventEmitter {
  /**
   * Create an AgentProcess.
   * @param {Object} props - Properties
   * @param {string} props.agentId - Agent identifier
   * @param {import('node:child_process').ChildProcess} props.process - Child process
   * @param {string} props.workingDir - Working directory
   */
  constructor({ agentId, process, workingDir }) {
    super();
    /** @type {string} */
    this.agentId = agentId;
    /** @type {import('node:child_process').ChildProcess} */
    this.process = process;
    /** @type {string} */
    this.workingDir = workingDir;
    /** @type {Date} */
    this.startedAt = new Date();
    /** @type {string[]} */
    this.outputLines = [];
    /** @type {string[]} */
    this.errorLines = [];
    /** @private @type {boolean} */
    this._terminated = false;

    this._setupOutputHandlers();
  }

  /**
   * Get the process ID.
   * @returns {number|null}
   */
  get pid() {
    return this.process?.pid ?? null;
  }

  /**
   * Check if the process is running.
   * @returns {boolean}
   */
  get isRunning() {
    return (
      this.process !== null &&
      !this.process.killed &&
      this.process.exitCode === null &&
      !this._terminated
    );
  }

  /**
   * Get the exit code.
   * @returns {number|null}
   */
  get returnCode() {
    return this.process?.exitCode ?? null;
  }

  /**
   * Setup stdout/stderr handlers.
   * @private
   */
  _setupOutputHandlers() {
    this.process.stdout?.on('data', (data) => {
      const lines = data.toString().split('\n').filter(Boolean);
      this.outputLines.push(...lines);
      for (const line of lines) {
        this.emit('stdout', line);
      }
    });

    this.process.stderr?.on('data', (data) => {
      const lines = data.toString().split('\n').filter(Boolean);
      this.errorLines.push(...lines);
      for (const line of lines) {
        this.emit('stderr', line);
      }
    });

    this.process.on('exit', (code, signal) => {
      this._terminated = true;
      this.emit('exit', code, signal);
    });

    this.process.on('error', (error) => {
      this.emit('error', error);
    });
  }

  /**
   * Send input to the process stdin.
   * @param {string} text - Text to send
   * @returns {boolean} True if sent successfully
   */
  sendInput(text) {
    if (this.process?.stdin && !this.process.stdin.destroyed) {
      this.process.stdin.write(text + '\n');
      return true;
    }
    return false;
  }

  /**
   * Get all stdout output as a single string.
   * @returns {string}
   */
  getOutput() {
    return this.outputLines.join('\n');
  }

  /**
   * Get all stderr output as a single string.
   * @returns {string}
   */
  getErrors() {
    return this.errorLines.join('\n');
  }

  /**
   * Get the last N lines of output.
   * @param {number} [n=10] - Number of lines
   * @returns {string[]}
   */
  getLastLines(n = 10) {
    return this.outputLines.slice(-n);
  }

  /**
   * Gracefully terminate the process.
   * @param {number} [timeoutMs=5000] - Timeout before force kill
   * @returns {Promise<void>}
   */
  async terminate(timeoutMs = 5000) {
    if (!this.process || this._terminated) return;

    return new Promise((resolvePromise) => {
      const timeout = setTimeout(() => {
        console.warn(`[Process] ${this.agentId} didn't terminate, killing`);
        this.process.kill('SIGKILL');
        this._terminated = true;
        resolvePromise();
      }, timeoutMs);

      this.process.once('exit', () => {
        clearTimeout(timeout);
        this._terminated = true;
        resolvePromise();
      });

      // Try graceful termination first
      this.process.kill('SIGTERM');
    });
  }

  /**
   * Force kill the process.
   * @returns {void}
   */
  kill() {
    if (this.process && !this._terminated) {
      this.process.kill('SIGKILL');
      this._terminated = true;
    }
  }
}

/**
 * Manages terminal processes for agents.
 */
export class TerminalManager {
  /**
   * Create a TerminalManager.
   * @param {string} baseDir - Base directory for sandboxes
   */
  constructor(baseDir) {
    /** @type {string} */
    this.baseDir = resolve(baseDir);
    /** @private @type {Map<string, AgentProcess>} */
    this._processes = new Map();
  }

  /**
   * Spawn a Claude Code CLI agent.
   *
   * @param {Object} props - Properties
   * @param {string} props.agentId - Agent identifier
   * @param {string} props.prompt - Initial prompt for the agent
   * @param {string} [props.workingDir] - Working directory (defaults to sandbox)
   * @param {boolean} [props.dangerouslySkipPermissions=true] - Skip permission prompts
   * @param {Object} [props.env] - Additional environment variables
   * @returns {Promise<AgentProcess>}
   */
  async spawnClaudeAgent({
    agentId,
    prompt,
    workingDir,
    dangerouslySkipPermissions = true,
    env = {},
  }) {
    const cwd = workingDir ?? resolve(this.baseDir, 'sandbox', agentId);
    await mkdir(cwd, { recursive: true });

    const args = ['--print'];
    if (dangerouslySkipPermissions) {
      args.push('--dangerously-skip-permissions');
    }
    args.push('-p', prompt);

    console.log(`[TerminalManager] Spawning agent ${agentId} in ${cwd}`);

    const childProcess = spawn('claude', args, {
      cwd,
      stdio: ['pipe', 'pipe', 'pipe'],
      env: { ...process.env, ...env },
    });

    const agentProcess = new AgentProcess({
      agentId,
      process: childProcess,
      workingDir: cwd,
    });

    // Setup logging
    agentProcess.on('stdout', (line) => {
      console.log(`[${agentId}] ${line}`);
    });

    agentProcess.on('stderr', (line) => {
      console.error(`[${agentId}] ERROR: ${line}`);
    });

    agentProcess.on('exit', (code) => {
      console.log(`[${agentId}] Exited with code ${code}`);
    });

    this._processes.set(agentId, agentProcess);
    return agentProcess;
  }

  /**
   * Spawn a generic command.
   *
   * @param {Object} props - Properties
   * @param {string} props.agentId - Identifier for this process
   * @param {string} props.command - Command to run
   * @param {string[]} [props.args=[]] - Command arguments
   * @param {string} [props.workingDir] - Working directory
   * @param {Object} [props.env] - Additional environment variables
   * @returns {Promise<AgentProcess>}
   */
  async spawnCommand({ agentId, command, args = [], workingDir, env = {} }) {
    const cwd = workingDir ?? this.baseDir;
    await mkdir(cwd, { recursive: true });

    console.log(`[TerminalManager] Spawning command ${command} as ${agentId}`);

    const childProcess = spawn(command, args, {
      cwd,
      stdio: ['pipe', 'pipe', 'pipe'],
      env: { ...process.env, ...env },
    });

    const agentProcess = new AgentProcess({
      agentId,
      process: childProcess,
      workingDir: cwd,
    });

    this._processes.set(agentId, agentProcess);
    return agentProcess;
  }

  /**
   * Get a process by agent ID.
   * @param {string} agentId - Agent identifier
   * @returns {AgentProcess|undefined}
   */
  getProcess(agentId) {
    return this._processes.get(agentId);
  }

  /**
   * Check if an agent process is running.
   * @param {string} agentId - Agent identifier
   * @returns {boolean}
   */
  isRunning(agentId) {
    const process = this._processes.get(agentId);
    return process?.isRunning ?? false;
  }

  /**
   * Terminate a specific agent process.
   * @param {string} agentId - Agent identifier
   * @param {number} [timeoutMs=5000] - Timeout before force kill
   * @returns {Promise<void>}
   */
  async terminate(agentId, timeoutMs = 5000) {
    const agentProcess = this._processes.get(agentId);
    if (agentProcess) {
      await agentProcess.terminate(timeoutMs);
      this._processes.delete(agentId);
    }
  }

  /**
   * Terminate all processes.
   * @param {number} [timeoutMs=5000] - Timeout before force kill
   * @returns {Promise<void>}
   */
  async terminateAll(timeoutMs = 5000) {
    const promises = [];
    for (const agentProcess of this._processes.values()) {
      promises.push(agentProcess.terminate(timeoutMs));
    }
    await Promise.all(promises);
    this._processes.clear();
  }

  /**
   * Get all agent IDs.
   * @returns {string[]}
   */
  getAgentIds() {
    return Array.from(this._processes.keys());
  }

  /**
   * Get count of running processes.
   * @returns {number}
   */
  getRunningCount() {
    let count = 0;
    for (const process of this._processes.values()) {
      if (process.isRunning) count++;
    }
    return count;
  }

  /**
   * Send input to a specific agent.
   * @param {string} agentId - Agent identifier
   * @param {string} text - Text to send
   * @returns {boolean} True if sent
   */
  sendInput(agentId, text) {
    const process = this._processes.get(agentId);
    return process?.sendInput(text) ?? false;
  }
}
